DSEGOG-100 tooltip warning message over search button if record count > limit

sam-glendenning opened the pull request:
## Description
This adds a warning message in the form of a tooltip over both the Search button and the filter dialogue Apply button if we detect the new search will return more records than the limit set in the settings file.

This is done through:
- creating a new `recordLimitWarning` setting in `operationsgateway-settings.json` to define the maximum number of records before this warning is displayed (set to -1 for unlimited)
- using an `incomingRecordCount` query to get the `count` of records in the new search before actually fetching those records
- if the count is greater than the limit set, a tooltip is displayed over the Search or Apply button (whichever is handling the search at that moment)
- the search is then only "confirmed" and initiated when the user clicks the button for a second time

There's had to be a significant rework to get this all to occur. ~~I've not yet found a way to transfer the incoming record count into the actual record count itself, meaning two requests for the count occur for each declared search. This isn't ideal.~~

I've also exposed a problem where there are two methods of searching (through the search components and the filter dialogue) but each only takes into account its own localised parameter changes. What this means is, if I change the shot number range in search, then add a filter and hit Apply, only the new filter parameters have been applied to the search while the shot number range is present in the shot number search component. So we'll need to find a solution to this down the line.

~~I would have liked to have found a way to not display the tooltip if it had already been displayed for a particular set of search parameters but was unable to find a reliable way of doing this.~~

Currently, the `recordLimitWarning` setting is mandatory and given a positive integer for an actual limit setting and -1 to denote unlimited but I can change this to make the setting optional instead.

I attempted some e2e tests for the tooltips displaying and the count request being made but I think the two count requests occurring has got in the way of this and I couldn't get very far with it. Fortunately, unit tests cover tooltip tests quite nicely.

So overall, this could use some improvement.

## Testing instructions
Tests should hopefully pass, it's just whether or not we want this specific implementation.

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage
- [ ] {more steps here}

## Agile board tracking
Closes DSEGOG-100

Commit:
DSEGOG-100 add new setting, check incoming count before search query

Adding a new recordLimitWarning setting to operationsgateway-settings.json. When set to an integer, any attempt to fetch more records than this number displays a popup warning message. If this is set to -1, no limit is imposed.

This is done through a new useQuery hook called useIncomingRecordCount. This is called with the potential search parameters and a count is fetched first. If the count is less than the limit or the user presses continue on the popup window, the search is actually performed and the records are fetched. Otherwise, the search is not performed.

Currently, this is only happening for the search bar, will make it happen for filters soon

Commit:
DSEGOG-100 adding warning for filters

Popup warning is now triggered if a filters search fetches over the limit in records

Commit:
Merge branch 'develop' into DSEGOG-100-data-limit-warning

Commit:
DSEGOG-100 improving readability

Commit:
DSEGOG-100 switching from popup to tooltip

Tooltip displays over Search button. Search does not initiate until second button click

Commit:
DSEGOG-100 adding warning tooltip to filter search

Commit:
DSEGOG-100 updating index tests

Accounts for new recordLimitWarning setting

Commit:
DSEGOG-100 adding test for tooltip display

Also checking search doesn't initiate until the second button click

Commit:
DSEGOG-100 fixing bug in record limit checker

Commit:
DSEGOG-100 adding tooltip test for filter dialogue

Also extending jest timeouts to account for slow user.type

Commit:
DSEGOG-100 adjusting e2e settings

Commit:
Merge branch 'develop' into DSEGOG-100-data-limit-warning

Commit:
DSEGOG-100 updating searchBar unit test

Factoring in maxShots addition

Commit:
DSEGOG-100 adding records api tests

Accounts for new useIncomingRecordCount function

Commit:
DSEGOG-100 updating configSlice tests

To account for new loadRecordLimitWarningSetting action

Commit:
DSEGOG-100 bumping coverage

Commit:
DSEGOG-100 fixing e2e tests

The search e2e tests need the recordLimitWarning to be set to -1 (hence the stubbed settings response) and an additional wait for an extra count request being made (I'm not sure if this is the incomingRecordCount query or a bug). But for some reason, the filtering e2e tests need none of this to pass, despite using the same functionality. This needs to be investigated

Commit:
DSEGOG-100 adding unit test for new state action

sam-glendenning commented:
The search e2e tests need the recordLimitWarning to be set to -1 (hence the stubbed settings response) and an additional wait for an extra count request being made (I'm not sure if this is the incomingRecordCount query or a bug). But for some reason, the filtering e2e tests need none of this to pass, despite using the same functionality. This needs to be investigated

louise-davies commented:
@sam-glendenning could perhaps using `initialData` and the query cache work for getting the count (see https://react-query-v3.tanstack.com/guides/initial-query-data#initial-data-from-cache)? If this doesn't work, then if we know that `incomingRecordCount` will already be calculated before any use of `recordCount`, perhaps we remove the `recordCount` query, and instead just populate `recordCount` from the query cache

louise-davies commented:
Also, good spot on the search vs filters issue. I don't think it's unreasonable behaviour to enforce that users need to click `Search`/`Apply` in order for their changes to be reflected so I'm not sure whether it is a "bug" as such that needs fixing.

sam-glendenning commented:
> @sam-glendenning could perhaps using `initialData` and the query cache work for getting the count (see https://react-query-v3.tanstack.com/guides/initial-query-data#initial-data-from-cache)? If this doesn't work, then if we know that `incomingRecordCount` will already be calculated before any use of `recordCount`, perhaps we remove the `recordCount` query, and instead just populate `recordCount` from the query cache

Ah good, this was the sort of thing I was looking for. I'll have a go at this

louise-davies reviewed (commented):
Did you manage to have a look at using the query cache? Submitting this as a review so it stops pinging me :)

sam-glendenning commented:
Sorry, I'd forgotten I'd committed to doing something more here. I'll mark it as draft until I need further review

Commit:
Merge branch 'develop' into DSEGOG-100-data-limit-warning

Commit:
DSEGOG-100 using initial data from incomingRecordCount

This eliminates the second, pointless query to the /count endpoint on the same parameters as before

Commit:
DSEGOG-100 adding unit test for query cache fetching

Adding a test to make sure that we use the earlier cached result from useIncomingRecordCount if we have one. Basically ensuring we only make one call to /records/count

Commit:
Merge branch 'develop' into DSEGOG-100-data-limit-warning

Commit:
DSEGOG-100 fixing timestamp format

Conflicts with new timestamp format on backend

sam-glendenning commented:
@louise-davies I've implemented your solution to reuse the cached data from `useIncomingRecordCount` in `useRecordCount` and it works well, nice fix!

I'd recommend giving this another review but this time in the context of how it runs in the browser. The most glaring problem at the moment is that the tooltip warning displays on searches we've already done (this can also mean that the user can perfrm a large search, get the results, then hit search over and over without changing any search parameters and the tooltip will display over and over).

Is there any way we can tell if a cached result is being used and therefore not show the tooltip if this happens? I am aware of an `isPreviousData` boolean returned from useQuery but I don't think this quite lines up with what I'm after because that only seems to compare the last query we ran, not all of them (and even then I couldn't seem to get it working quite right)

louise-davies commented:
@sam-glendenning I believe I have a solution to this using `queryClient.getQueriesData` to check if the `records` query has data for the given filters/search params combo the user is attempting to search. See gif below to see if you're happy with the new behaviour: (in the demo, I search with min shot number 1, change to another param and prompt the message without searching, switch back to 1 and click search without the message, and then go back to 2 and see the message again)

![OG-search-warning](https://user-images.githubusercontent.com/22525471/208075206-c6e3e614-707e-4015-9771-2b57410dc155.gif)

sam-glendenning commented:
That looks like what I'm after! Would you be able to commit that? I can write the tests, fix merge conflicts etc.

louise-davies committed:
DSEGOG-100 only show data limit warning on non-cached searches

louise-davies commented:
@sam-glendenning committed :)

Commit:
Merge branch 'develop' into DSEGOG-100-data-limit-warning

Commit:
DSEGOG-100 attempting to test tooltip reads query cache

In the middle of trying to test that the tooltip will not display if we launch a search on a previously executed query. Not managed it so far so committing what I have in case it's salvageable down the line

Commit:
DSEGOG-100 adding e2e test for tooltips

This makes sure the tooltip only appears in the right moments

Commit:
Revert "DSEGOG-100 attempting to test tooltip reads query cache"

This reverts commit 4d3fc61e3e67ee6dab5f1e5bbefb579fc33541bf.

Commit:
Merge branch 'develop' into DSEGOG-100-data-limit-warning

sam-glendenning commented:
@louise-davies I've got this ready for another review. The main functionality I was trying to test was:
1. initiate a search from date X and check it displays tooltip first
2. initiate a search from date Y and check it displays tooltip first
3. initiate a search from date X and check no tooltip displays, instead launching the search straight away (since the user has already seen the tooltip for this search)

Attempts to test this functionality in RTL failed (I was trying to `setQueryData` on a test query client but it didn't seem to pick up on any existing query data during the test). However, I did manage to test this in an e2e test so I would say this is adequate. If you want to unit test this, you can check what I committed in https://github.com/ral-facilities/operationsgateway/pull/149/commits/4d3fc61e3e67ee6dab5f1e5bbefb579fc33541bf which I have since reverted since it's no longer relevant.

louise-davies committed:
DSEGOG-100 refactor dateTime error handling
React was complaining that we were setting state
in the input render function, so refactor so we don't do that

louise-davies committed:
DSEGOG-100 add test for cached requests not triggering limit warning
Sam did the groundwork for this, I just fixed it

louise-davies commented on cypress/integration/table/search.spec.ts:
```diff
@@ -199,130 +455,164 @@ describe('Search', () => {
-        const lte: string = timestampRange['$lte'];
-        expect(gte).equal(expectedFromDateString);
-        expect(lte).equal(expectedToDateString);
+  describe.only('with record limits', () => {
```
Remove this only

louise-davies committed:
DSEGOG-100 add cached = no warning unit test to filterDialogue

louise-davies committed:
DSEGOG-100 remove only & add eslint rule to catch future accidental only's

louise-davies reviewed (approved):
I've just added/fixed tests from Sam's stuff, and I reviewed Sam's stuff before Christmas and it was all working good so I'll approve
